Update Golang to v1.15.0-rc.1

justaugustus opened the pull request:
**What type of PR is this?**

/kind cleanup
/area dependency release-eng
/sig release

**What this PR does / why we need it**:

Update Golang to v1.15.0

Depends on fixes in earlier PRs (which must also be included if go1.15 support is backported):
- [x] https://github.com/kubernetes/kubernetes/pull/93425 (int->string casts)
- [x] https://github.com/kubernetes/kubernetes/pull/93426 (add subjectAltName to serving certificates)

Followups (which must also be done in earlier releases if go1.15 support is backported):
- [x] update CI images to go1.15rc1 - https://github.com/kubernetes/test-infra/pull/18475
- [ ] update CI jobs to use go1.15.rc1 images (bot PR)
- [x] update lint-dependencies to stop exempting golang.org/x/... deps from minimal version selection per https://github.com/kubernetes/kubernetes/pull/93264#issuecomment-663933730 - https://github.com/kubernetes/kubernetes/pull/93603
- [ ] once 1.15.0 is released, set `minimum_go_version=go1.15` in `hack/lib/golang.sh`

Signed-off-by: Stephen Augustus <saugustus@vmware.com>

cc: @kubernetes/release-engineering @kubernetes/release-team-leads @kubernetes/sig-scalability-feature-requests @liggitt @dims 
Tracking issue: https://github.com/kubernetes/release/issues/1421

**Special notes for your reviewer**:

This is testing against a pre-release version of go1.15 (`go1.15rc1`).
~Expect a slew of errors, as various parts of our code are not configured to accept Go pre-release versions.~

~Also, there will definitely be bazel toolchain + rules_go errors.~

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
- Kubernetes is now built with golang 1.15.0-rc.1.
- The deprecated, legacy behavior of treating the CommonName field on X.509 serving certificates as a host name when no Subject Alternative Names are present is now disabled by default. It can be temporarily re-enabled by adding the value x509ignoreCN=0 to the GODEBUG environment variable.
```

justaugustus committed:
[go1.15] dependencies.yaml: Update to go1.15rc1

Signed-off-by: Stephen Augustus <saugustus@vmware.com>

justaugustus committed:
[go1.15] build: Update to k/repo-infra@v0.0.10 (supports go1.15rc1)

Signed-off-by: Stephen Augustus <saugustus@vmware.com>

justaugustus commented:
/test pull-kubernetes-verify

justaugustus commented:
Alright, bazel has beaten me for the day.

I have a PR (https://github.com/kubernetes/repo-infra/pull/197) running against k/repo-infra, which:
- pulls in the go1.15beta1 pre-releases
- bumps gazelle to v0.21.1

This PR targets that k/repo-infra branch/commit, but I'm pretty consistently hitting errors like this:

```console
 package conflict error: k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/api/equality: package imports k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/apis/meta/v1
	  was compiled with: //staging/src/k8s.io/apimachinery/pkg/apis/meta/v1:go_default_library
	but was linked with: //staging/src/k8s.io/apimachinery/pkg/apis/meta/v1:go_default_test
This sometimes happens when an external test (package ending with _test)
imports a package that imports the library being tested. This is not supported.
See https://github.com/bazelbuild/rules_go/issues/1877. 
```

Examples:
- https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/93264/pull-kubernetes-bazel-test/1285464783922925569
- https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/93264/pull-kubernetes-bazel-build/1285464783939702785

This seems to match behavior in https://github.com/bazelbuild/bazel-gazelle/issues/839, but I'm unsure how to proceed.

Any thoughts here, @jayconrod @ixdy @fejta @mikedanese?

jayconrod commented:
Thanks for the heads up. I'll investigate these failures to see what should be done on the rules_go side to mitigate or fix them.

To explain what's going on:

When `go test` builds a test binary, it divides the tests into two packages: the internal tests are compiled together with the library under test, and the external tests are compiled separately. The external test may import the internal test package (using the library's import path), and it may depend on test-only symbols. In order to avoid conflicts, every package the external test imports that transitively imports the library under test must be recompiled to import the internal test package instead.

The `go_test` rule has never actually done the recompilation part of this, since it scales poorly and is somewhat difficult to implement in Bazel. `go test` has an easier time because it can read source files during analysis, so it can decide specifically when this needs to happen. We only saw a small number of link errors before (bazelbuild/rules_go#1877), and it seemed like it would be better just to work around the problem. 

Go 1.15 has a new linker which is stricter than the old linker. It checks that each linked package was compiled against packages with the same export data as are passed to the linker. So in Go 1.15, this pattern becomes an error much more frequently.

The linker conflict errors are somewhat baffling, so I've added a more descriptive error to the link wrapper, which is showing up here.

In any case, I'll look into these failures and see if something can be done in rules_go to fix them reasonably soon.

liggitt commented:
https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/93264/pull-kubernetes-node-e2e/1285396468525436928 flake is #93210

fedebongio commented:
/remove-sig api-machinery

justaugustus commented:
@jayconrod -- Thanks for the response! We'll hang tight until there's a rules_go update :)

justaugustus committed:
build/dependencies.yaml: Group related dependency entries

Signed-off-by: Stephen Augustus <saugustus@vmware.com>

justaugustus commented:
(sorry for the noise; just rebasing out the yesterday's bazel yak shave)

jayconrod commented:
bazelbuild/rules_go#2579 should fix the `go_test` linker conflicts. It simulates what `go test` does. Please try it out. If it works, I'll backport it and tag a point release.

justaugustus commented:
@jayconrod -- I'm trying to hack that in with https://github.com/kubernetes/repo-infra/pull/197/commits/56161307d45d6ab30dc17983ba653b9ad9b8022d (https://github.com/kubernetes/repo-infra/pull/197), but am running into some issues that I haven't seen before. I'll ping you in that PR.

justaugustus commented:
@jayconrod -- UPDATE :)

https://github.com/kubernetes/kubernetes/pull/93264/commits/d457623fdbcb0b28d9359b807da0a7cea50a8f1c tests against https://github.com/kubernetes/repo-infra/pull/197, which pulls in https://github.com/bazelbuild/rules_go/pull/2579 and that seems to have fixed the link issues.

Compare with https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/93264/pull-kubernetes-bazel-test/1286113168686321664 (which is a real test failure)

https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/93264/pull-kubernetes-bazel-test/1285464783922925569 (which contains link errors):

```
 INFO: From GoLink staging/src/k8s.io/apimachinery/pkg/conversion/queryparams/linux_amd64_race_stripped/go_default_test:
package conflict error: k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime: package imports k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/conversion/queryparams
	  was compiled with: //staging/src/k8s.io/apimachinery/pkg/conversion/queryparams:go_default_library
	but was linked with: //staging/src/k8s.io/apimachinery/pkg/conversion/queryparams:go_default_test
This sometimes happens when an external test (package ending with _test)
imports a package that imports the library being tested. This is not supported.
See https://github.com/bazelbuild/rules_go/issues/1877.
This will be an error in the future.
ERROR: /home/prow/go/src/github.com/kubernetes/kubernetes/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/BUILD:9:1: GoLink staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/linux_amd64_race_stripped/go_default_test failed (Exit 1): builder failed: error executing command 
```

So I think based on this, you should be in the clear to cut a release with this fix in it! :)

In general, we're on rules_go v0.22.z. Would you be able to backport this fix to that release stream as well?

Thanks so much for your help!

justaugustus committed:
[go1.15] Remove client support for darwin/386

Removed in go1.15 (GH/golang/go/issues/37610).

Signed-off-by: Stephen Augustus <saugustus@vmware.com>

justaugustus commented:
https://github.com/kubernetes/kubernetes/pull/93264/commits/d14c0290699b74ad20e3710a92c2809faf07d9d8 removes darwin/386 client support (ref: https://github.com/golang/go/issues/37610)

https://github.com/kubernetes/kubernetes/pull/93264/commits/2d27742b3d52058ab4fc5b758b49b20e7738475f updates golang/x/** deps (thanks for the pointers in https://github.com/kubernetes/kubernetes/pull/93317, @liggitt)

justaugustus commented:
Assigning Jordan so the Triage team doesn't get mad at me (ref: https://kubernetes.slack.com/archives/C2C40FMNF/p1595515477184100?thread_ts=1595515402.183700&cid=C2C40FMNF)
/assign @liggitt

justaugustus commented:
Opened https://github.com/kubernetes/repo-infra/pull/201 to pull in the new [rules_go@v0.22.9](https://github.com/bazelbuild/rules_go/releases/tag/v0.22.9).
Thanks @jayconrod!

justaugustus commented:
Before I rebase to pull in the new rules_go from repo-infra, I want to capture what passing/failing as of https://github.com/kubernetes/kubernetes/pull/93264/commits/2d27742b3d52058ab4fc5b758b49b20e7738475f:

### Passing

- pull-kubernetes-bazel-build
- pull-kubernetes-cross
- pull-kubernetes-dependencies
- pull-kubernetes-e2e-gce-100-performance
- pull-kubernetes-e2e-gce-device-plugin-gpu
- pull-kubernetes-files-remake
- pull-kubernetes-integration
- pull-kubernetes-kubemark-e2e-gce-big
- pull-kubernetes-node-e2e
- pull-kubernetes-typecheck
- pull-kubernetes-verify

### Failing

#### [pull-kubernetes-bazel-test](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/93264/pull-kubernetes-bazel-test/1286142738688905217)

`//staging/src/k8s.io/kube-aggregator/pkg/apiserver:go_default_test`:

```console
exec ${PAGER:-/usr/bin/less} "$0" || exit 1
Executing tests from //staging/src/k8s.io/kube-aggregator/pkg/apiserver:go_default_test
-----------------------------------------------------------------------------
E0723 03:44:56.271929      24 handler_proxy.go:146] error resolving test-ns/bad-service: unresolveable
2020-07-23 03:44:57.069468 I | http: TLS handshake error from 127.0.0.1:56050: remote error: tls: bad certificate
2020-07-23 03:44:57.138186 I | http: TLS handshake error from 127.0.0.1:35758: remote error: tls: bad certificate
2020-07-23 03:45:02.171992 I | http: TLS handshake error from 127.0.0.1:57064: remote error: tls: bad certificate
2020-07-23 03:45:02.188637 I | http: TLS handshake error from 127.0.0.1:57068: remote error: tls: bad certificate
--- FAIL: TestProxyCertReload (0.04s)
    handler_proxy_test.go:671: Expected status code 200 but got 503
FAIL
```

`//staging/src/k8s.io/apimachinery/pkg/util/proxy:go_default_test`:

```console
exec ${PAGER:-/usr/bin/less} "$0" || exit 1
Executing tests from //staging/src/k8s.io/apimachinery/pkg/util/proxy:go_default_test
-----------------------------------------------------------------------------
2020/07/23 03:41:49 http: TLS handshake error from 127.0.0.1:49084: remote error: tls: bad certificate
2020/07/23 03:41:49 http: TLS handshake error from 127.0.0.1:54974: remote error: tls: bad certificate
2020/07/23 03:41:49 http: TLS handshake error from 127.0.0.1:45848: remote error: tls: bad certificate
2020/07/23 03:41:49 http: TLS handshake error from 127.0.0.1:58188: remote error: tls: bad certificate
E0723 03:41:53.762929      22 upgradeaware.go:312] Proxy upgrade error: invalid upgrade response: status code 200
E0723 03:41:53.766616      22 upgradeaware.go:312] Proxy upgrade error: invalid upgrade response: status code 300
E0723 03:41:53.772717      22 upgradeaware.go:312] Proxy upgrade error: invalid upgrade response: status code 307
E0723 03:41:53.775363      22 upgradeaware.go:312] Proxy upgrade error: invalid upgrade response: status code 200
E0723 03:41:53.778080      22 upgradeaware.go:312] Proxy upgrade error: invalid upgrade response: status code 300
E0723 03:41:53.781143      22 upgradeaware.go:312] Proxy upgrade error: invalid upgrade response: status code 302
E0723 03:41:53.786694      22 upgradeaware.go:312] Proxy upgrade error: invalid upgrade response: status code 307
--- FAIL: TestProxyRequestContentLengthAndTransferEncoding (0.07s)
    upgradeaware_test.go:924: content-length + identity transfer-encoding: Did not get successful response: HTTP/1.1 501 Not Implemented
        Content-Type: text/plain; charset=utf-8
        Connection: close
        
        Unsupported transfer encoding
FAIL
```

#### kind e2es

- [pull-kubernetes-e2e-kind](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/93264/pull-kubernetes-e2e-kind/1286142738705682432)
- [pull-kubernetes-conformance-kind-ga-only-parallel](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/93264/pull-kubernetes-conformance-kind-ga-only-parallel/1286142738701488135)
- [pull-kubernetes-conformance-kind-ipv6-parallel](https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/93264/pull-kubernetes-conformance-kind-ipv6-parallel/1286142738701488128/)

```console
 Building node image in: /tmp/kind-node-image606705008
ERROR: error building node image: failed to copy build artifact: lstat /home/prow/go/src/k8s.io/kubernetes/bazel-bin/cmd/kubeadm/linux_amd64_stripped/kubeadm: no such file or directory
Stack Trace: 
sigs.k8s.io/kind/pkg/errors.Wrap
	/src/pkg/errors/errors.go:39
sigs.k8s.io/kind/pkg/build/nodeimage.(*buildContext).populateBits
	/src/pkg/build/nodeimage/build_impl.go:79
sigs.k8s.io/kind/pkg/build/nodeimage.(*buildContext).Build
	/src/pkg/build/nodeimage/build_impl.go:58
sigs.k8s.io/kind/pkg/build/nodeimage.Build
	/src/pkg/build/nodeimage/build.go:68
sigs.k8s.io/kind/pkg/cmd/kind/build/nodeimage.runE
	/src/pkg/cmd/kind/build/nodeimage/nodeimage.go:72
sigs.k8s.io/kind/pkg/cmd/kind/build/nodeimage.NewCommand.func1
	/src/pkg/cmd/kind/build/nodeimage/nodeimage.go:46
github.com/spf13/cobra.(*Command).execute
	/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:842
github.com/spf13/cobra.(*Command).ExecuteC
	/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:950
github.com/spf13/cobra.(*Command).Execute
	/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:887
sigs.k8s.io/kind/cmd/kind/app.Run
	/src/cmd/kind/app/main.go:53
sigs.k8s.io/kind/cmd/kind/app.Main
	/src/cmd/kind/app/main.go:35
main.main
	/src/main.go:25
runtime.main
	/usr/local/go/src/runtime/proc.go:204
runtime.goexit
	/usr/local/go/src/runtime/asm_amd64.s:1374
wrapper.sh] [TEST] Test Command exit code: 1
wrapper.sh] [CLEANUP] Cleaning up after Docker in Docker ...
Stopping Docker: dockerProgram process in pidfile '/var/run/docker-ssd.pid', 1 process(es), refused to die.
wrapper.sh] [CLEANUP] Done cleaning up after Docker in Docker. 
```

#### [pull-kubernetes-e2e-gce](https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/93264/pull-kubernetes-e2e-gce/1286142738701488131/)

Multiple webhook timeouts:

```console
test/e2e/framework/framework.go:597
Jul 23 04:11:46.505: waiting for webhook configuration to be ready
Unexpected error:
    <*errors.errorString | 0xc00020e200>: {
        s: "timed out waiting for the condition",
    }
    timed out waiting for the condition
occurred
test/e2e/apimachinery/webhook.go:1749
```

#### [pull-kubernetes-e2e-gce-ubuntu-containerd](https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/93264/pull-kubernetes-e2e-gce-ubuntu-containerd/1286142738701488134/)

Multiple webhook timeouts:

```console
test/e2e/framework/framework.go:597
Jul 23 03:57:14.190: waiting for webhook configuration to be ready
Unexpected error:
    <*errors.errorString | 0xc00026c1f0>: {
        s: "timed out waiting for the condition",
    }
    timed out waiting for the condition
occurred
test/e2e/apimachinery/webhook.go:909
```

liggitt commented:
for pull-kubernetes-integration + pull-kubernetes-verify, how did you make those tests run with go1.15? the CI job configuration determines the go version, right?

liggitt commented:
https://tip.golang.org/doc/go1.15#net/http:

> support for the "identity" Transfer-Encoding was dropped

justaugustus committed:
[go1.15] apimachinery/pkg/util/proxy: Drop identity transfer-encoding

ref: https://tip.golang.org/doc/go1.15#net/http

Parsing is now stricter as a hardening measure against request smuggling
attacks: non-ASCII white space is no longer trimmed like SP and HTAB,
and support for the "identity" Transfer-Encoding was dropped.

Signed-off-by: Stephen Augustus <saugustus@vmware.com>

justaugustus commented:
> for pull-kubernetes-integration + pull-kubernetes-verify, how did you make those tests run with go1.15? the CI job configuration determines the go version, right?

@liggitt -- It depends? :tm: 

For tests using bazel, https://github.com/kubernetes/kubernetes/pull/93264/commits/3d2b1056da4e85657c45589b43aa52c4f6beb25c hacks in https://github.com/kubernetes/repo-infra/pull/197, which pulls in go1.15beta1 go_sdk.

For tests depending on cross, we've built a kube-cross:v1.15-beta.1-1 image: https://github.com/kubernetes/release/pull/1420

For tests otherwise depending on a go image, we're referencing the golang:1.15beta1 image.

AFAIK, there isn't a canary integration or verify test, so elements of those tests will still be go1.14.6.

justaugustus commented:
> https://tip.golang.org/doc/go1.15#net/http:
> 
> > support for the "identity" Transfer-Encoding was dropped

Fixed in https://github.com/kubernetes/kubernetes/pull/93264/commits/75eb162b3ab0a30d8a71d7ca4467666fe999f4ae.

liggitt commented:
TestProxyCertReload and webhook timeouts are probably due to serving certificates that are no longer considered valid, see https://github.com/golang/go/issues/39568

Fix in the last commit of https://github.com/liggitt/kubernetes/commits/test-certs

liggitt commented:
> AFAIK, there isn't a canary integration or verify test, so elements of those tests will still be go1.14.6.

yeah, integration was the one I was wondering about... it exercises a ton of stuff that could be impacted by https://github.com/kubernetes/kubernetes/pull/93264#issuecomment-663200033

justaugustus commented:
> TestProxyCertReload and webhook timeouts are probably due to serving certificates that are no longer considered valid, see [golang/go#39568](https://github.com/golang/go/issues/39568)
> 
> Fix in the last commit of https://github.com/liggitt/kubernetes/commits/test-certs

Thanks! Pulled in as https://github.com/kubernetes/kubernetes/pull/93264/commits/7c45c8e980795903480b5a76bb4a5c8aaebbf66e.

Worklog for generating the cert --> https://kubernetes.slack.com/archives/C2C40FMNF/p1595534523188700?thread_ts=1595270953.060000&cid=C2C40FMNF

> > AFAIK, there isn't a canary integration or verify test, so elements of those tests will still be go1.14.6.
> 
> yeah, integration was the one I was wondering about... it exercises a ton of stuff that could be impacted by [#93264 (comment)](https://github.com/kubernetes/kubernetes/pull/93264#issuecomment-663200033)

I vaguely anticipated this and added a `go-canary` variant for kubekins-e2e in https://github.com/kubernetes/test-infra/pull/18328, which is currently at go1.15beta1.

I can PR some go canary presubmits to test both in tandem or maybe we need something like similar to `build/root/.bazelversion` for go, so we can switch versions if it exists...

justaugustus commented:
A few of the previously failing tests are starting to pass:
- pull-kubernetes-bazel-build
- pull-kubernetes-conformance-kind-ga-only-parallel
- pull-kubernetes-conformance-kind-ipv6-parallel

**cautious optimism swells**

justaugustus commented:
/test pull-kubernetes-e2e-kind
/test pull-kubernetes-e2e-gce
/test pull-kubernetes-e2e-gce-100-performance

justaugustus commented:
/test pull-kubernetes-bazel-test

justaugustus commented:
I've opened https://github.com/kubernetes/repo-infra/pull/202, which allows for overriding the Go version in bazel and will be needed here for removing the hackery that currently points to my fork of k/repo-infra.

---

(Some of the tests seem to not be reporting in the GH context.)
/retest

justaugustus commented:
For posterity before I rebase this to include https://github.com/kubernetes/repo-infra/releases/tag/v0.0.9:

![Screenshot from 2020-07-23 19-56-47](https://user-images.githubusercontent.com/567897/88349696-ccfbfa00-cd1e-11ea-9f66-f883a468f3ce.png)
![Screenshot from 2020-07-23 19-57-05](https://user-images.githubusercontent.com/567897/88349703-d08f8100-cd1e-11ea-9a76-08870f7d1e73.png)

BenTheElder commented:
https://github.com/kubernetes/kubernetes/issues/92937#issuecomment-663303165
/retest

justaugustus commented:
/test pull-kubernetes-dependencies

justaugustus commented:
Bah. Thwarted by `/hack/verify-hack-tools.sh`, which to be fair, I just found out about :sweat_smile: (ref: https://github.com/kubernetes/kubernetes/pull/90252).

The remaining failures are in:
- `pull-kubernetes-verify` (needed `./hack/update-hack-tools.sh`)
- `pull-kubernetes-e2e-gce-100-performance` (timeout)

More tomorrow.

justaugustus commented:
/test pull-kubernetes-e2e-kind

justaugustus commented:
/test pull-kubernetes-e2e-gce

justaugustus commented:
FYI @kubernetes/sig-scalability -- this PR is go1.15beta1, in case you need/want something to test against.

justaugustus commented:
/test pull-kubernetes-e2e-kind

justaugustus commented:
/retest

liggitt commented:
Are you sure the unit tests are running with go1.15? Got a report of test failures in https://github.com/kubernetes/kubernetes/issues/93407#issue-664916762

BenTheElder commented:
In presubmit they should be using whatever WORKSPACE uses, however it wouldn't be the first time we had skew between that and make test / go test

justaugustus commented:
Rebased and updated k/repo-infra to pull in go1.15rc1.
Going to see some failures here as the kube-cross image hasn't been promoted yet. (ref: https://github.com/kubernetes/release/pull/1437)

justaugustus commented:
/retest
(kube-cross:v1.15.0-rc.1-1 has been promoted: https://github.com/kubernetes/k8s.io/pull/1063)

justaugustus commented:
/test pull-kubernetes-e2e-gce-100-performance

justaugustus commented:
https://github.com/kubernetes/kubernetes/pull/93426 and https://github.com/kubernetes/kubernetes/pull/93425 have been rebased in.

The [verify failure](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/93264/pull-kubernetes-verify/1286825780222889985) suggested running `./hack/update-bazel.sh` to fix the diff, but that consistently came up clean on my machine.

Example verify-bazel:

```
$ time make verify WHAT="bazel"
Verifying verify-bazel.sh

+++ Running case: verify.bazel 
+++ working dir: /home/augustus/go/src/k8s.io/kubernetes
+++ command: bash "hack/make-rules/../../hack/verify-bazel.sh"
go: downloading github.com/bazelbuild/bazel-gazelle v0.21.1
go: downloading github.com/bazelbuild/buildtools v0.0.0-20200228172928-c9d9e342afdb
go: downloading golang.org/x/sync v0.0.0-20200625203802-6e8e738ad208
go: downloading github.com/pelletier/go-toml v1.2.0
go: downloading github.com/bmatcuk/doublestar v1.2.2
go: downloading golang.org/x/tools v0.0.0-20200710042808-f1c4188a97a1
go: downloading github.com/pmezard/go-difflib v1.0.0
go: downloading github.com/golang/protobuf v1.3.4
go: downloading k8s.io/repo-infra v0.0.10
go: downloading k8s.io/klog/v2 v2.3.0
go: downloading github.com/go-logr/logr v0.2.0
+++ exit code: 0
SUCCESS  verify-bazel.sh	92s

real	1m32.509s
user	0m20.744s
sys	0m15.847s
```

Running `bazel clean --expunge`, followed by `./hack/update-vendor.sh` (which also updates bazel), seems to have fixed it.

After that, I ran:

```console
$ for foo in crypto net oauth2 sys time tools; do hack/pin-dependency.sh golang.org/x/$foo master; done
Running: go get golang.org/x/crypto@master
Resolved to golang.org/x/crypto@v0.0.0-20200709230013-948cd5f35899
Running: go mod edit -replace golang.org/x/crypto=golang.org/x/crypto@v0.0.0-20200709230013-948cd5f35899

Run hack/update-vendor.sh to rebuild the vendor directory
Running: go get golang.org/x/net@master
Resolved to golang.org/x/net@v0.0.0-20200707034311-ab3426394381
Running: go mod edit -replace golang.org/x/net=golang.org/x/net@v0.0.0-20200707034311-ab3426394381

Run hack/update-vendor.sh to rebuild the vendor directory
Running: go get golang.org/x/oauth2@master
Resolved to golang.org/x/oauth2@v0.0.0-20200107190931-bf48bf16ab8d
Running: go mod edit -replace golang.org/x/oauth2=golang.org/x/oauth2@v0.0.0-20200107190931-bf48bf16ab8d

Run hack/update-vendor.sh to rebuild the vendor directory
Running: go get golang.org/x/sys@master
Resolved to golang.org/x/sys@v0.0.0-20200724161237-0e2f3a69832c
Running: go mod edit -replace golang.org/x/sys=golang.org/x/sys@v0.0.0-20200724161237-0e2f3a69832c

Run hack/update-vendor.sh to rebuild the vendor directory
Running: go get golang.org/x/time@master
Resolved to golang.org/x/time@v0.0.0-20200630173020-3af7569d3a1e
Running: go mod edit -replace golang.org/x/time=golang.org/x/time@v0.0.0-20200630173020-3af7569d3a1e

Run hack/update-vendor.sh to rebuild the vendor directory
Running: go get golang.org/x/tools@master
Resolved to golang.org/x/tools@v0.0.0-20200724172932-b5fc9d354d99
Running: go mod edit -replace golang.org/x/tools=golang.org/x/tools@v0.0.0-20200724172932-b5fc9d354d99

Run hack/update-vendor.sh to rebuild the vendor directory

$ hack/lint-dependencies.sh --all | grep "  hack" | bash

$ hack/update-vendor.sh
+++ [0725 02:40:41] logfile at /tmp/update-vendor.PHOJ/update-vendor.log
+++ [0725 02:40:41] go.mod: update staging references
+++ [0725 02:40:48] go.mod: propagate to staging modules
+++ [0725 02:40:49] go.mod: sorting staging modules
+++ [0725 02:41:02] go.mod: tidying
+++ [0725 02:41:14] go.mod: adding generated comments
+++ [0725 02:41:14] vendor: running 'go mod vendor'
+++ [0725 02:41:16] vendor: updating BUILD files
+++ [0725 02:41:21] vendor: updating vendor/LICENSES
+++ [0725 02:41:28] vendor: creating OWNERS file
+++ [0725 02:41:28] NOTE: don't forget to handle vendor/* files that were added or removed
```

and added that to https://github.com/kubernetes/kubernetes/pull/93264/commits/fc1843af80aa93171e5285be8b97f21b12d24be0.

This should be ready for approval now.

I've also opened https://github.com/kubernetes/test-infra/pull/18475 to build the kubekins-e2e variants using go1.15rc1.

justaugustus commented:
/test pull-kubernetes-conformance-kind-ga-only-parallel
(pod pending timeout)

dims commented:
dependencies look good. the new `github.com/yuin/goldmark` we pull in references, but not actually use the code.

Looks like the other 2 PRs merged already as well.

/lgtm

justaugustus commented:
(FYI -- this will need a final rebase once https://github.com/kubernetes/kubernetes/pull/93356 merges, as the entries in `build/dependencies.yaml` get shuffled around a bit.)

justaugustus commented:
Rebased to pull in https://github.com/kubernetes/kubernetes/pull/93356.
Now we should be ready for approval. :)

hakman commented:
/retest

liggitt commented on LICENSES/vendor/golang.org/x/mod/LICENSE:
```diff
@@ -1,4 +1,4 @@
-= vendor/golang.org/x/crypto licensed under: =
+= vendor/golang.org/x/mod licensed under: =
 
```
This is strange… does this mean we're missing the crypto license now even though we're still vendoring code from it?

liggitt replied:
nm, this got dropped from `hack/tools/LICENSES/vendor/golang.org/x/crypto/LICENSE` but is still alive and well in `LICENSES/vendor/golang.org/x/crypto/LICENSE`

liggitt commented:
do we have a run of the integration job with go1.15?

justaugustus committed:
[go1.15] tools: Update to k/repo-infra@v0.0.10 (supports go1.15rc1)

Signed-off-by: Stephen Augustus <saugustus@vmware.com>

liggitt committed:
generated: hack/update-hack-tools.sh && hack/update-vendor.sh

liggitt committed:
[go1.15] go.mod: update to go1.15

liggitt committed:
generated: hack/update-vendor.sh

liggitt committed:
[go1.15] update golang.org/x/... deps to 1.15 versions

hack/pin-dependency.sh golang.org/x/crypto  release-branch.go1.15
hack/pin-dependency.sh golang.org/x/mod     release-branch.go1.15
hack/pin-dependency.sh golang.org/x/net     release-branch.go1.15
hack/pin-dependency.sh golang.org/x/text    v0.3.3
hack/pin-dependency.sh golang.org/x/tools   release-branch.go1.15
hack/pin-dependency.sh golang.org/x/xerrors release-branch.go1.15

Did not pin golang.org/x/sys to release-branch.go1.15 because we require a newer level

liggitt committed:
generated: hack/update-vendor.sh

liggitt committed:
[go1.15] fix pin-dependency.sh to only use module info

liggitt commented:
suggestions:
* move the "build/dependencies.yaml: Group related dependency entries" commit ahead of the version change commits
* squash the beta1 and rc1 commits together
* do each of the following as a distinct commit to make it clearer where the changes come from:
  * update repo-tools
  * run update-hack-tools + update-vendor (updates BUILD files, mostly)
  * update to "go 1.15" directive in go.mod files
  * run update-vendor (updates modules.txt, mostly)
  * only update golang.org dependencies that have release-branch.go1.15 branches:
    ```
    hack/pin-dependency.sh golang.org/x/crypto  release-branch.go1.15
    hack/pin-dependency.sh golang.org/x/mod     release-branch.go1.15
    hack/pin-dependency.sh golang.org/x/net     release-branch.go1.15
    hack/pin-dependency.sh golang.org/x/text    release-branch.go1.15
    hack/pin-dependency.sh golang.org/x/tools   release-branch.go1.15
    hack/pin-dependency.sh golang.org/x/xerrors release-branch.go1.15
    ```

  * run update-vendor

I pushed https://github.com/liggitt/kubernetes/commits/go115 with those adjustments done

liggitt commented:
setting `minimum_go_version=go1.15` in `hack/lib/golang.sh` would ensure that the integration and verify jobs were running go1.15... shouldn't we do that and get a green CI run before merging?

liggitt commented:
looks like the green test-integration run is still using go1.14:

> `+++ [0725 10:54:57] Checking kubectl version
Client Version: version.Info{Major:"1", Minor:"20+", GitVersion:"v1.20.0-alpha.0.380+1ebe43b85cc0dc", GitCommit:"1ebe43b85cc0dced1a2db3319acc887976b2100b", GitTreeState:"clean", BuildDate:"2020-07-25T09:58:18Z", GoVersion:"go1.14.6", Compiler:"gc", Platform:"linux/amd64"}`
> `Server Version: version.Info{Major:"1", Minor:"20+", GitVersion:"v1.20.0-alpha.0.380+1ebe43b85cc0dc", GitCommit:"1ebe43b85cc0dced1a2db3319acc887976b2100b", GitTreeState:"clean", BuildDate:"2020-07-25T09:58:18Z", GoVersion:"go1.14.6", Compiler:"gc", Platform:"linux/amd64"}`

justaugustus commented:
Reset to @liggitt's reorganized branch and rebased from master.

I think next steps:
1. Let the tests do their thing
2. Jordan is doing a local run of `make test-cmd` and `make test-integration` on 1.15rc1
3. If successful, merge
4. Merge the kubekins-e2e variants updates: https://github.com/kubernetes/test-infra/pull/18475
5. I'll send a note to k-dev about go1.15
6. Prow bump to pick up the new Golang versions in CI
7. Open another PR to `minimum_go_version` to go1.15 and carry that, along with any other upstream Go changes, until 1.15.0 is officially released**

The Prow bump should wait until Monday, because there are a few consumers of the kubekins-e2e `master` variant image and they should be informed ahead of any breakages.

** Updating the `minimum_go_version` in this PR going to cause failures because the CI images are on go1.14.6

liggitt commented:
> 2\. Jordan is doing a local run of `make test-cmd` and `make test-integration` on 1.15rc1

make test-cmd was successful

justaugustus commented:
/test pull-kubernetes-bazel-test

justaugustus commented:
/test pull-kubernetes-e2e-kind

liggitt commented:
unit test failure:

> 957/957 tests passed!
>
> --- FAIL: TestRecycleSlices/1_to_create_and_1_to_delete_for_each_IP_family (0.00s)
>
> ERROR: /home/prow/go/src/github.com/kubernetes/kubernetes/pkg/security/podsecuritypolicy/seccomp/BUILD:20:1: GoLink pkg/security/podsecuritypolicy/seccomp/linux_amd64_race_stripped/go_default_test failed (Exit 34): java.io.IOException: io.grpc.StatusRuntimeException: UNAVAILABLE: 502:Bad Gateway

uhhh...

justaugustus commented:
/test pull-kubernetes-e2e-gci-gce-ipvs

justaugustus commented:
/test pull-kubernetes-e2e-gce

liggitt commented:
> 2. Jordan is doing a local run of `make test-cmd` and `make test-integration` on 1.15rc1

test-integration is ... less so... will dig later

```
FAIL	k8s.io/kubernetes/test/integration/apiserver	161.781s
FAIL	k8s.io/kubernetes/test/integration/apiserver/admissionwebhook	617.137s
FAIL	k8s.io/kubernetes/test/integration/apiserver/apply	199.580s
ok  	k8s.io/kubernetes/test/integration/apiserver/certreload	52.658s
ok  	k8s.io/kubernetes/test/integration/apiserver/flowcontrol	24.246s
ok  	k8s.io/kubernetes/test/integration/apiserver/podlogs	16.559s
FAIL	k8s.io/kubernetes/test/integration/auth	167.708s
FAIL	k8s.io/kubernetes/test/integration/certificates	155.913s
FAIL	k8s.io/kubernetes/test/integration/client	137.219s
ok  	k8s.io/kubernetes/test/integration/configmap	13.083s
ok  	k8s.io/kubernetes/test/integration/cronjob	35.822s
FAIL	k8s.io/kubernetes/test/integration/daemonset	286.666s
ok  	k8s.io/kubernetes/test/integration/defaulttolerationseconds	8.108s
FAIL	k8s.io/kubernetes/test/integration/deployment	93.824s
ok  	k8s.io/kubernetes/test/integration/disruption	11.465s
ok  	k8s.io/kubernetes/test/integration/dryrun	39.748s
ok  	k8s.io/kubernetes/test/integration/endpointslice	9.908s
FAIL	k8s.io/kubernetes/test/integration/etcd	122.867s
ok  	k8s.io/kubernetes/test/integration/events	8.817s
ok  	k8s.io/kubernetes/test/integration/evictions	21.148s
FAIL	k8s.io/kubernetes/test/integration/examples	40.909s
FAIL	k8s.io/kubernetes/test/integration/garbagecollector	186.295s
ok  	k8s.io/kubernetes/test/integration/ipamperf	5.181s
ok  	k8s.io/kubernetes/test/integration/kubelet	13.840s
ok  	k8s.io/kubernetes/test/integration/master	427.410s
ok  	k8s.io/kubernetes/test/integration/metrics	11.869s
ok  	k8s.io/kubernetes/test/integration/namespace	61.178s
ok  	k8s.io/kubernetes/test/integration/node	51.924s
ok  	k8s.io/kubernetes/test/integration/objectmeta	8.050s
ok  	k8s.io/kubernetes/test/integration/openshift	11.707s
ok  	k8s.io/kubernetes/test/integration/pods	44.298s
ok  	k8s.io/kubernetes/test/integration/quota	17.060s
ok  	k8s.io/kubernetes/test/integration/replicaset	104.390s
ok  	k8s.io/kubernetes/test/integration/replicationcontroller	94.733s
ok  	k8s.io/kubernetes/test/integration/scale	10.363s
ok  	k8s.io/kubernetes/test/integration/scheduler	466.389s
ok  	k8s.io/kubernetes/test/integration/scheduler_perf	3.507s
ok  	k8s.io/kubernetes/test/integration/secrets	11.750s
ok  	k8s.io/kubernetes/test/integration/serviceaccount	53.805s
ok  	k8s.io/kubernetes/test/integration/serving	45.782s
ok  	k8s.io/kubernetes/test/integration/statefulset	24.650s
ok  	k8s.io/kubernetes/test/integration/storageclasses	9.801s
ok  	k8s.io/kubernetes/test/integration/tls	11.214s
FAIL	k8s.io/kubernetes/test/integration/ttlcontroller	38.855s
??? k8s.io/kubernetes/test/integration/volume
??? k8s.io/kubernetes/test/integration/volumescheduling
```

BenTheElder commented:
Maybe I missed something but I checked out this PR and `make test-integration` passed with `go version go1.15rc1 linux/amd64` AFAICT.

liggitt commented:
Good…? Weird, let me try on my linux machine

FiloSottile commented:
> * only update golang.org dependencies that have release-branch.go1.15 branches:
>   ```
>   hack/pin-dependency.sh golang.org/x/crypto  release-branch.go1.15
>   hack/pin-dependency.sh golang.org/x/mod     release-branch.go1.15
>   hack/pin-dependency.sh golang.org/x/net     release-branch.go1.15
>   hack/pin-dependency.sh golang.org/x/text    release-branch.go1.15
>   hack/pin-dependency.sh golang.org/x/tools   release-branch.go1.15
>   hack/pin-dependency.sh golang.org/x/xerrors release-branch.go1.15
>   ```

Please don't use the release branches of the golang.org subrepos. They are not what you might think they are: we make release branches to track the version that is vendored _into a Go release_, not the version that is compatible with a Go release. These release branches are not documented and not supported, and if there is a security vulnerability in a package that is not vendored into the standard library, they will not get the update. You should treat golang.org modules like any other module.

justaugustus commented:
Thanks @FiloSottile! So you're saying we'd be better suited targeting the `master` branch for the golang/x/... tools instead?

FiloSottile commented:
> Thanks @FiloSottile! So you're saying we'd be better suited targeting the `master` branch for the golang/x/... tools instead?

Yes, if that's your policy for other dependencies. If you are using Go modules, the standard thing to track would be `latest`, which translates to `master` if there are no semver tags, and to the latest tag if there is one.

liggitt commented:
/lgtm
/approve

followups:
- [ ] update CI images to go1.15rc1 - https://github.com/kubernetes/test-infra/pull/18475
- [ ] update CI jobs to use go1.15.rc1 images (bot PR)
- [ ] update lint-dependencies to stop exempting golang.org/x/... deps from minimal version selection per https://github.com/kubernetes/kubernetes/pull/93264#issuecomment-663933730 - @liggitt
- [ ] once 1.15.0 is released, set `minimum_go_version=go1.15` in `hack/lib/golang.sh`

justaugustus commented:
> added [followups](https://github.com/kubernetes/kubernetes/pull/93264#issue-453810890) to the description

Thanks @liggitt! Full tracking issue is here and I'll get the follow-ups added to it: https://github.com/kubernetes/release/issues/1421

liggitt commented:
```
--- FAIL: TestCustomResourceCascadingDeletion
garbage_collector_test.go:960: expected dependent "dependentqdrg7" to be deleted
```
which [isn't normally a flaky test](https://storage.googleapis.com/k8s-gubernator/triage/index.html?pr=1&test=TestCustomResourceCascadingDeletion)

/retest

dims commented:
@FiloSottile thanks for the guidance!
